src/auth.rs

```diff
@@ -0,0 +1,209 @@
+#[derive(Clone, Debug)]
+pub struct Authentication(u8);
+
+impl Default for Authentication {
+    fn default() -> Self {
+        Self::none()
+    }
+}
+
+impl Authentication {
+    /// Disable all authentication schemes. This is the default.
+    pub const fn none() -> Self {
+        Authentication(0)
+    }
+
+    /// Enable all available authentication schemes.
+    pub const fn all() -> Self {
+        #[allow(unused_mut)]
+        let mut all = Self::basic().0 | Self::digest().0;
+
+        #[cfg(feature = "spnego")]
+        {
+            all |= Self::negotiate().0;
+        }
+
+        Authentication(all)
+    }
+
+    /// HTTP Basic authentication.
+    ///
+    /// This authentication scheme sends the user name and password over the
+    /// network in plain text. Avoid using this scheme without TLS as the
+    /// credentials can be easily captured otherwise.
+    pub const fn basic() -> Self {
+        Authentication(0b0001)
+    }
+
+    /// HTTP Digest authentication.
+    ///
+    /// Digest authentication is defined in RFC 2617 and is a more secure way to
+    /// do authentication over public networks than the regular old-fashioned
+    /// Basic method.
+    pub const fn digest() -> Self {
+        Authentication(0b0010)
+    }
+
+    /// HTTP Negotiate (SPNEGO) authentication.
+    ///
+    /// Negotiate authentication is defined in RFC 4559 and is the most secure
+    /// way to perform authentication over HTTP.
+    ///
+    /// You need to build libcurl with a suitable GSS-API library or SSPI on
+    /// Windows for this to work. This is automatic when binding to curl
+    /// statically, otherwise it depends on how your system curl is configured.
+    #[cfg(feature = "spnego")]
+    pub const fn negotiate() -> Self {
+        Authentication(0b0100)
+    }
+
+    #[inline]
+    fn contains(&self, other: Self) -> bool {
+        (self.0 & other.0) == other.0
+    }
+
+    fn as_auth(&self) -> curl::easy::Auth {
+        let mut auth = curl::easy::Auth::new();
+
+        if self.contains(Authentication::basic()) {
+            auth.basic(true);
+        }
+
+        if self.contains(Authentication::digest()) {
+            auth.digest(true);
+        }
+
+        #[cfg(feature = "spnego")]
+        {
+            if self.contains(Authentication::negotiate()) {
+                auth.gssnegotiate(true);
+            }
+        }
+
+        auth
+    }
+}
+
+impl BitOr for Authentication {
+    type Output = Self;
+
+    fn bitor(self, other: Self) -> Self {
+        Authentication(self.0 | other.0)
+    }
+}
+
+impl BitOrAssign for Authentication {
+    fn bitor_assign(&mut self, rhs: Self) {
+        *self = self.clone() | rhs;
+    }
+}
+
+impl SetOpt for Authentication {
+    fn set_opt<H>(&self, easy: &mut curl::easy::Easy2<H>) -> Result<(), curl::Error> {
+        #[cfg(feature = "spnego")]
+        {
+            if self.contains(Authentication::negotiate()) {
+                easy.username(":")?;
```

Comment:
This works on Linux, but does not work on Windows (presumably something different in the SSPI codepath?). Instead, set to the empty string (or whatever), but you must also explicity set `password()` to something (i.e. empty string) in order to allow negotiate to work properly on Windows.

Reply:
Interesting. In Windows CI it appears that the `Authorization` header is being populated with _something_ (see the Windows-specific test in `tests/auth.rs`). But other than that I have no way to test this feature to see if it works. :cry: Does the empty string break it on Linux (meaning we must have platform-specific codepaths)?

Also, is this documented somewhere? I couldn't find much on this.

Reply:
From looking at both the cURL code and the documentation, I can tell that *something* needs to be set in order for Negotiate to work, but in some places, it is referred to as the username and in others as the password. On the command line, you would normally pass `-u username:password` (or more specifically for negotiate, "`-u :`"), which I suspect is how you ended up with `":"`.

But to specifically answer your question: As far as I can tell, both Windows and Linux don't care what you set for the username/passwords when doing Negotiate. For some reason it works on Linux if you *only* set username, but Windows wants *both* username and password. So you shouldn't need any platform-specific code, just set both username and password to the empty string on all platforms (at least that's how I did my testing, and that worked).

(of course, the fun thing is that for Negotiate, both the username and password are always ignored, setting them just "turns on" cURL's ability to do *any* authentication)

Reply:
OK, that makes sense. I'll update the code to reflect that.